splitgraph/hooks/data_source/fdw.py

```diff
@@ -299,6 +344,119 @@ def _load(self, schema: str, tables: Optional[TableInfo] = None):
+        query = (
+            SQL("SELECT ")
+            + SQL(",").join(Identifier(p) for p in cursor_fields)
+            + SQL(" FROM {}.{} ORDER BY ").format(Identifier(schema), Identifier(table))
+            + SQL(",").join(Identifier(p) + SQL(" DESC") for p in cursor_fields)
+        ) + SQL(" LIMIT 1")
+
+        return [str(r) for r in self.engine.run_sql(query, return_shape=ResultShape.ONE_MANY)]
+
+    def sync(
+        self,
+        repository: Repository,
+        image_hash: Optional[str] = None,
+        tables: Optional[TableInfo] = None,
+        use_state: bool = True,
+    ) -> str:
+
+        self._validate_table_params(tables)
+        tables = tables or self.tables
+
+        # Load ingestion state
+        base_image, new_image_hash = prepare_new_image(
+            repository, image_hash, comment=f"{self.get_name()} data load"
+        )
+        state = get_ingestion_state(repository, image_hash) if use_state else None
+        logging.info("Current ingestion state: %s", state)
+
+        # Set up a staging schema for the data
+        staging_schema = get_temporary_table_id()
+
+        with delete_schema_at_end(repository.object_engine, staging_schema):
+            repository.object_engine.delete_schema(staging_schema)
+            repository.object_engine.create_schema(staging_schema)
+            repository.commit_engines()
+
+            self._mount_and_copy(
+                staging_schema,
+                tables,
+                cursor_values=None if not state else state.get("cursor_values"),
+            )
+
+            logging.info("Storing tables as Splitgraph images")
+            for table_name in repository.object_engine.get_all_tables(staging_schema):
+                logging.info("Storing %s", table_name)
+                new_schema = repository.object_engine.get_full_table_schema(
+                    staging_schema, table_name
+                )
+
+                if base_image:
+                    try:
+                        current_schema = base_image.get_table(table_name).table_schema
+                        if current_schema != new_schema:
+                            raise AssertionError(
+                                "Schema for %s changed! Old: %s, new: %s"
+                                % (
+                                    table_name,
+                                    current_schema,
+                                    new_schema,
+                                )
+                            )
+                    except TableNotFoundError:
+                        pass
+
+                repository.objects.record_table_as_base(
+                    repository,
+                    table_name,
+                    new_image_hash,
+                    chunk_size=DEFAULT_CHUNK_SIZE,
+                    source_schema=staging_schema,
+                    source_table=table_name,
+                    table_schema=new_schema,
+                )
```

Comment:
There's some overlap between this code and the code in https://github.com/splitgraph/splitgraph/blob/master/splitgraph/ingestion/airbyte/data_source.py#L290-L329 but I don't know if there's a nice way to factor it out + when writeable LQ lands, we should be able to simplify this code anyway by writing into the LQ checkout.

Reply:
Yeah, makes sense.